### ISSUE TITLE
Problem with loading behaviour when doing SSR and PersistGate

### ISSUE THREAD

**Issue by rajatbarman** [rt2zz/redux-persist#1008]
Hey guys,

Great solve for the persistence problem, although I have an issue.

`
<PersistGate loading="loading" persistor={persistor}>
    <MyEntireApp />
</PersistGate/>
`

When the app is rendered from the server, the react component tree is already created and the view is already rendered, but because of <PersistGate /> the whole tree is reconstructed as it has to render the loading text instead of the whole app while persist is syncing with the store and then again construct the whole app's component tree, which leads to a break in user experience.

Imagine this - 
Web app Rendered (from server) ->
Loading Text Rendered (when client JS comes into play and persistor starts hydrating) ->
Web app Re-rendered (after persistor is done hydrating the store)

A not so user friendly experience.

Basically this line in PersistGate creates problem when doing SSR  - 
`
   this.state.bootstrapped ? this.props.children : this.props.loading
`

**Comment by rajatbarman** [comment]
To those looking for solution.
```
persistor.subscribe(() => {
          /* Hydrate React components when persistor has synced with redux store */
          const { bootstrapped } = persistor.getState();

          if (bootstrapped) {
              ReactDOM.hydrate(
                  <MyEntireApp />,
                  document.getElementById("appOrWhatever")
            );
          }
        });
```

**Comment by pwnreza** [comment]
Hi there.
Thanks for your answer.
Didn't you have an Error similar to:
"Warning: Did not expect server HTML to contain a <div> in <div>."

I guess that because the Dom Tree is different without the PersistGate you get the Error while rehydrating.

**Comment by max8hine** [comment]
*redux-persist* do cause a lot of problem in SSR life-cycle of Gatsby.
could not load the persisted data correctly
and also [providing google/facebook crawling the Gatsby site](https://github.com/gatsbyjs/gatsby/issues/11245)
Any solution or progress for it?

**Comment by krunalp1993** [comment]
@rajatbarman Your solution worked for me! Thanks!

**Comment by MiKatre** [comment]
My solution:
https://stackoverflow.com/a/62501513/7292383

**Comment by alexpchin** [comment]
I'm using struggling to solve this when integrating `redux-persist` with Gatsby.

**Comment by ybarakhovskyi** [comment]
https://github.com/vercel/next.js/issues/8240#issuecomment-647699316 another solution from Next.JS community

**Comment by simplenotezy** [comment]
> To those looking for solution. Don't use `<PersistGate />` wrapper, instead wrap this logic over ReactDOM.hydrate
> 
> ```
> persistor.subscribe(() => {
>           /* Hydrate React components when persistor has synced with redux store */
>           const { bootstrapped } = persistor.getState();
> 
>           if (bootstrapped) {
>               ReactDOM.hydrate(
>                   <MyEntireApp />,
>                   document.getElementById("appOrWhatever")
>             );
>           }
>         });
> ```

@rajatbarman can you show where you put the code you provided? No clue where to put your snippet.

**Comment by 6lyxt** [comment]
bump
